docs/design/unified_sidecar_and_process_monitor.md

```diff
@@ -0,0 +1,118 @@
+
+During upgrades, we will have a sidecar container that we upgrade to the new version, with instructions to copy the binary into a shared volume. We will then update the configuration for fdb-kubernetes-monitor in the main container to tell it to use a newer version, which will cause it to use the version-scoped binary in the shared volume. We will include the sidecar container all the time in the initial work for this feature. In the future, we will explore an option to deploy it as an ephemeral container, once that feature is sufficiently stable in Kubernetes.
+
+fdb-kubernetes-monitor will log its activity to standard output, and will have an option to also log to a file.
+
+fdb-kubernetes-monitor will ship in an image called `foundationdb-kubernetes`. This image will be versioned with FoundationDB, and will include multiple client libraries, like the `foundationdb-kubernetes-sidecar` image currently does. The first `foundationdb-kubernetes` image we cut for a release will have the FoundationDB version as its tag. For subsequent updates to the image that do not involve a FoundationDB update, we will append a build number, of the format `-2`. For instance, we would have images in a series like `6.3.0`, `6.3.0-2`, `6.3.0-3`, and so on.
+
+In order to keep the new image lightweight, and allow the development team for the operator to easily manage the development on the monitor, we will write the fdb-kubernetes-monitor process in go.
+
+We will have an option in the cluster spec to use the new launcher, which will be off by default. We will have an entire major release of FoundationDB where we publish both the old sidecar image and the new kubernetes image, to aid with the transition. For any version of FoundationDB in that major release, the user will have the option of which launch strategy to use. Upon the next major release, we will stop publishing the old sidecar image, which means users will need to switch to the new launcher before that release.
+
+When the new launcher is used, both the main container and the sidecar container will run the same image, but the sidecar container will have a different set of arguments that cause it to only copy the fdbserver binary, where necessary, and to open up a listener. The main container will load its configuration file, launch the configured server processes, and open a listener. That listener will be an HTTP server that provides an API for checking the hashes of files and other local configuration, similar to the API the sidecar currently provides. However, the fdb-kubernetes-monitor API will be not cover all use cases of the sidecar API. It will provide the following API calls:
+
+* `get_current_configuration` - returns the current process configuration
+* `check_hash` - check the hash of a file inside its configuration directory
+* `ready` - returns an `OK` response
+* `substitutions` - returns the values for any environment variables that are used in the process configuration
```

Comment:
I would also add a prometheus `/metrics` endpoint.

Reply:
I think we should save that for a follow-up design, since it introduces more questions about the scope of the metrics we're going to expose through this API.

Reply:
Sounds good